go/internal/services/keys/verifier.go

```diff
@@ -0,0 +1,135 @@
+	Roles                 []string                                      // RBAC roles assigned to this key
+	Permissions           []string                                      // RBAC permissions assigned to this key
+	Valid                 bool                                          // Whether the key passed all validation checks
+	Status                KeyStatus                                     // The current validation status
+	AuthorizedWorkspaceID string                                        // The workspace ID this key is authorized for
+	RatelimitResults      map[string]RatelimitConfigAndResult           // Combined config and results for rate limits (name -> config+result)
+	isRootKey             bool                                          // Whether this is a root key (special handling)
+	session               *zen.Session                                  // The current request session
+	rateLimiter           ratelimit.Service                             // Rate limiting service
+	usageLimiter          usagelimiter.Service                          // Usage limiting service
+	rBAC                  *rbac.RBAC                                    // Role-based access control service
+	clickhouse            clickhouse.ClickHouse                         // Clickhouse for telemetry
+	logger                logging.Logger                                // Logger for verification operations
+	message               string                                        // Internal message for validation failures
+}
+
+// GetRatelimitConfigs returns the rate limit configurations (for internal use)
+func (k *KeyVerifier) GetRatelimitConfigs() map[string]db.KeyFindForVerificationRatelimit {
+	return k.ratelimitConfigs
+}
+
+// Verify performs key verification with the given options.
+// For root keys: returns fault errors for validation failures.
+// For normal keys: returns error only for system problems, check k.Valid and k.Status for validation results.
+func (k *KeyVerifier) Verify(ctx context.Context, opts ...VerifyOption) error {
+	// Skip verification if key is already invalid
+	if !k.Valid {
+		// For root keys, auto-return validation failures as fault errors
+		if k.isRootKey {
+			return k.ToFault()
+		}
+		return nil
+	}
+
+	// nolint:exhaustruct
+	config := &verifyConfig{}
+	for _, opt := range opts {
+		if err := opt(config); err != nil {
+			return err
+		}
+	}
+
+	var err error
+	if config.credits != nil {
+		err = k.withCredits(ctx, *config.credits)
+		if err != nil {
+			return err
+		}
+	}
+
+	if config.ipWhitelist {
+		err = k.withIPWhitelist()
+		if err != nil {
+			return err
+		}
+	}
+
+	if config.permissions != nil {
+		err = k.withPermissions(ctx, *config.permissions)
+		if err != nil {
+			return err
+		}
+	}
+
+	if config.apiID != nil {
+		k.WithApiID(*config.apiID)
+	}
+
+	err = k.withRateLimits(ctx, config.ratelimits)
+	if err != nil {
+		return err
+	}
+
+	k.clickhouse.BufferKeyVerification(schema.KeyVerificationRequestV1{
```

Comment:
this only runs when the key passed all verifications above, right? so we would only log valid verifications

Reply:
but would it log invalid regular keys? Say when they got ratelimited

Reply:
What it wouldn't log would be StatusWorkspaceDisabled / disabled keys for root keys since with root keys we would error out before. 

I think we should just move the verification insert out of the key verifier and just do it in the verify key endpoint itself or in the getRootKey one wdyt?